File: csharp/glp_link/seam/LinkOptions.cs
namespace GlpRuntime.Link.Seam;

/// <summary>
/// Per-link tunables passed to <see cref="ILinkTransport.ListenAsync"/> /
/// <see cref="ILinkTransport.ConnectAsync"/>. Every value is a tuning parameter,
/// NOT a correctness condition (architecture-context.md §5): defaults preserve the
/// GLP invariants; overrides only adjust resource bounds and timeouts.
/// </summary>
public sealed record LinkOptions
{
    /// <summary>The default for all fields.</summary>
    public static readonly LinkOptions Default = new();

    /// <summary>
    /// Bounded outbound window before the producer suspends (FR-025). Default
    /// <c>8</c>. Scheme-overridable BELOW the seam; the producer suspends (FCP
    /// suspension, not an unbounded buffer) so there is no OOM and no
    /// head-of-line blocking across independent links.
    /// </summary>
    public int BackpressureWindow { get; init; } = 8;

    /// <summary>
    /// Require transport-level encryption for inter-host links (FR-029). Default
    /// <c>true</c>: a plain inter-host link is refused (TLS/DTLS by default). A
    /// leaf MAY ignore this for an intrinsically in-process scheme
    /// (<see cref="LinkScheme.Loopback"/>).
    /// </summary>
    public bool RequireTls { get; init; } = true;

    /// <summary>
    /// Maximum transport frame size in bytes, or <c>null</c> for no leaf-imposed
    /// limit. Constrained leaves (CoAP ~1 KB, BLE GATT 20 B) set this so the
    /// framing layer (T021) fragments/reassembles under it (FR-022).
    /// </summary>
    public int? MaxFrameBytes { get; init; }

    /// <summary>
    /// Bounded silence after which the sublayer surfaces <c>tempFail</c> (FR-045,
    /// SC-010). A liveness tuning knob, not a correctness bound.
    /// </summary>
    public TimeSpan TempFailAfter { get; init; } = TimeSpan.FromSeconds(5);

    /// <summary>
    /// Bounded silence after which the sublayer gives up with <c>permFail</c>
    /// (FR-045). A deliberate, possibly-wrong give-up (FLP/two-generals): bounds
    /// blast radius, does not remove the wall.
    /// </summary>
    public TimeSpan PermFailAfter { get; init; } = TimeSpan.FromSeconds(30);

    /// <summary>Timeout for the initial listen/connect rendezvous.</summary>
    public TimeSpan ConnectTimeout { get; init; } = TimeSpan.FromSeconds(15);
}

File: glp_runtime/lib/link/seam/link_options.dart
import 'link_scheme.dart';

/// Per-link tunables passed to `ILinkTransport.listen` /
/// `ILinkTransport.connect`. Every value is a tuning parameter,
/// NOT a correctness condition (architecture-context.md §5): defaults preserve the
/// GLP invariants; overrides only adjust resource bounds and timeouts.
class LinkOptions {
  /// The default for all fields.
  static final LinkOptions default_ = LinkOptions();

  /// Bounded outbound window before the producer suspends (FR-025). Default
  /// `8`. Scheme-overridable BELOW the seam; the producer suspends (FCP
  /// suspension, not an unbounded buffer) so there is no OOM and no
  /// head-of-line blocking across independent links.
  final int backpressureWindow;

  /// Require transport-level encryption for inter-host links (FR-029). Default
  /// `true`: a plain inter-host link is refused (TLS/DTLS by default). A
  /// leaf MAY ignore this for an intrinsically in-process scheme
  /// ([LinkScheme.loopback]).
  final bool requireTls;

  /// Maximum transport frame size in bytes, or `null` for no leaf-imposed
  /// limit. Constrained leaves (CoAP ~1 KB, BLE GATT 20 B) set this so the
  /// framing layer (T021) fragments/reassembles under it (FR-022).
  final int? maxFrameBytes;

  /// Bounded silence after which the sublayer surfaces `tempFail` (FR-045,
  /// SC-010). A liveness tuning knob, not a correctness bound.
  final Duration tempFailAfter;

  /// Bounded silence after which the sublayer gives up with `permFail`
  /// (FR-045). A deliberate, possibly-wrong give-up (FLP/two-generals): bounds
  /// blast radius, does not remove the wall.
  final Duration permFailAfter;

  /// Timeout for the initial listen/connect rendezvous.
  final Duration connectTimeout;

  LinkOptions({
    this.backpressureWindow = 8,
    this.requireTls = true,
    this.maxFrameBytes,
    this.tempFailAfter = const Duration(seconds: 5),
    this.permFailAfter = const Duration(seconds: 30),
    this.connectTimeout = const Duration(seconds: 15),
  });

  @override
  bool operator ==(Object other) =>
      other is LinkOptions &&
      backpressureWindow == other.backpressureWindow &&
      requireTls == other.requireTls &&
      maxFrameBytes == other.maxFrameBytes &&
      tempFailAfter == other.tempFailAfter &&
      permFailAfter == other.permFailAfter &&
      connectTimeout == other.connectTimeout;

  @override
  int get hashCode => Object.hash(backpressureWindow, requireTls, maxFrameBytes,
      tempFailAfter, permFailAfter, connectTimeout);
}
